scheds/rust/scx_tickless/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,696 @@
+		dispatched = true;
+
+		bpf_task_release(p);
+
+		break;
+	}
+
+	if (dispatched && !is_primary_cpu(cpu))
+		scx_bpf_kick_cpu(cpu, SCX_KICK_IDLE);
+
+	return dispatched;
+}
+
+/*
+ * Consume tasks from the shared queue and distribute them evenly across
+ * the available CPUs.
+ *
+ * If @do_idle_smt is true, consider only full-idle SMT cores.
+ *
+ * Return true if there are still some free dispatch slots available, false
+ * otherwise.
+ */
+static bool dispatch_all_cpus(bool do_idle_smt)
+{
+	const struct cpumask *online_cpumask, *idle_smtmask, *idle_cpumask;
+	s32 cpu;
+
+	online_cpumask = scx_bpf_get_online_cpumask();
+	idle_smtmask = scx_bpf_get_idle_smtmask();
+	idle_cpumask = scx_bpf_get_idle_cpumask();
+
+	if (do_idle_smt && bpf_cpumask_empty(idle_smtmask))
+		goto out_put_cpumask;
+
+	if (!do_idle_smt && bpf_cpumask_empty(idle_cpumask))
+		goto out_put_cpumask;
+
+	bpf_for(cpu, 0, nr_cpu_ids) {
+		/*
+		 * Do not distribute tasks to offline or primary CPUs.
+		 */
+		if (!bpf_cpumask_test_cpu(cpu, online_cpumask) || is_primary_cpu(cpu))
+			continue;
+
+		/*
+		 * Skip if we want only full-idle SMT cores and the SMT is
+		 * busy.
+		 */
+		if (do_idle_smt && !bpf_cpumask_test_cpu(cpu, idle_smtmask))
+			continue;
+
+		/*
+		 * Skip busy CPUs.
+		 */
+		if (!bpf_cpumask_test_cpu(cpu, idle_cpumask))
+			continue;
+
+		/*
+		 * Try to dispatch a task that was using this CPU first, if
+		 * @prefer_same_cpu is enabled.
+		 */
+		if (!prefer_same_cpu || !dispatch_cpu(cpu, true))
+			dispatch_cpu(cpu, false);
+
+		/*
+		 * Stop dispatching tasks if we're out of dispatch slots.
+		 */
+		if (!scx_bpf_dispatch_nr_slots())
+			break;
+	}
+
+out_put_cpumask:
+	scx_bpf_put_cpumask(idle_smtmask);
+	scx_bpf_put_cpumask(idle_cpumask);
+	scx_bpf_put_cpumask(online_cpumask);
+
+	if (!scx_bpf_dispatch_nr_slots()) {
+		scx_bpf_kick_cpu(pick_primary_cpu(), SCX_KICK_PREEMPT);
+		return false;
+	}
+
+	return true;
+}
+
+void BPF_STRUCT_OPS(tickless_dispatch, s32 cpu, struct task_struct *prev)
+{
+	if (is_primary_cpu(cpu)) {
+		if (smt_enabled && !dispatch_all_cpus(true))
+			return;
+
+		if (!dispatch_all_cpus(false))
+			return;
+
+		/*
+		 * If there are still tasks in the shared DSQ after
+		 * distributing them to the tickless CPUs, start to consume
+		 * them also from the primary CPUs.
+		 */
+		if (scx_bpf_dsq_move_to_local(SHARED_DSQ)) {
+			__sync_fetch_and_add(&nr_fallback_dispatches, 1);
+			return;
+		}
+	} else {
+		if (scx_bpf_dsq_move_to_local(SHARED_DSQ)) {
+			__sync_fetch_and_add(&nr_fallback_dispatches, 1);
+			return;
+		}
+
+		/*
+		 * This tickless CPU is about to go idle: kick and preempt
+		 * a primary CPU so that it can dispatch a task here.
+		 *
+		 * We need to send a preemption IPI since primary CPUs may
+		 * also run a task with infinite time slice if all the
+		 * non-primary CPUs were busy and a task was dispatched
+		 * there.
+		 */
+		scx_bpf_kick_cpu(pick_primary_cpu(), SCX_KICK_PREEMPT);
```

Comment:
Wouldn't it be cheaper for !primary cpu to dispatch directly if it's about to go idle anyway?

Reply:
Actually we're already dispatching directly, we've just called `scx_bpf_dsq_move_to_local(SHARED_DSQ)`. So we didn't consume any task from SHARED_DSQ it means either that SHARED_DSQ is empty (and in this case it's pointless to kick a primary CPU) or none of remaining tasks in the SHARED_DSQ can be consumed on this CPU. But also in this case a primary CPU won't do much, because if there are CPUs that can consume tasks from the SHARED_DSQ they can run them from their ops.dispatch().

In conclusion, I think we can just remove this `scx_bpf_kick_cpu()` and let the CPU go idle without doing anything special.